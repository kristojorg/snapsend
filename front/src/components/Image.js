//@flow
import * as React from 'react';
import type { Image as ImageType } from '../types';
import styled from 'styled-components';
import T from '../components/T';
import CheckCircle from 'material-ui-icons/CheckCircle';
import { withTheme } from 'material-ui/styles';
import Delete from 'material-ui-icons/Delete';
import IconButton from 'material-ui/IconButton';

type Props = {
  img?: ImageType,
  toggle?: number => void,
  delete?: number => void,
  status?: 'DOWNLOADING' | 'REVIEWING' | 'EDITING',
};

type State = {
  hover: boolean,
};

const MAX_SIZE = 300;

export default class ImageComponent extends React.Component<Props, State> {
  state = {
    hover: false,
  };

  mouseEnter = () => this.setState({ hover: true });
  mouseLeave = () => this.setState({ hover: false });

  render() {
    const { img, toggle, delete: deleteImage, status } = this.props;
    const { hover } = this.state;
    if (!img) {
      return (
        <Outer>
          <Inner>
            <T variant="body1">Uploading image</T>
          </Inner>
        </Outer>
      );
    }
    const arr = img.url.split('/');
    const handle = arr[arr.length - 1];
    const src = `https://cdn.filestackcontent.com/resize=w:${MAX_SIZE},h:${MAX_SIZE}/output=f:jpg,c:true,q:80/${handle}`;
    return (
      <Outer
        onMouseEnter={this.mouseEnter}
        onMouseLeave={this.mouseLeave}
        onClick={toggle}
        selected={img.selected}
      >
        <Check selected={img.selected} />
        <Image src={src} selected={img.selected} />
        {status === 'EDITING' &&
          hover && (
            <IconOverlay onClick={deleteImage}>
              <Delete style={{ fontSize: 50, color: 'white' }} />
            </IconOverlay>
          )}
      </Outer>
    );
  }
}

const Outer = styled.div`
  position: relative;
  padding-top: 100%;
  background: rgba(0, 0, 0, 0.6);
  transition: 20ms ease-in-out;
  transform: ${({ selected }) => (selected ? 'scale(0.9)' : 'scale(1)')};
  cursor: pointer;
`;

const Inner = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: papayawhip;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const Image = styled.img`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  object-fit: contain;
`;

const IconOverlay = styled.div`
  position: absolute;
  top: 0;
  width: 100%;
  height: 100%;
  left: 0;
  background: rgba(0, 0, 0, 0.5);
  display: flex;
  justify-content: center;
  align-items: center;
`;

const Check = withTheme()(styled(CheckCircle)`
  position: absolute;
  top: -10px;
  opacity: ${({ selected }) => (selected ? 1 : 0)};
  transition: 20ms ease-in-out;
  right: -10px;
  color: ${({ theme }) => theme.palette.secondary.main};
  transform: scale(1.5);
  background: white;
  border-radius: 100%;
`);
