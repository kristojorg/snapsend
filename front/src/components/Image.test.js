import React from 'react';
import Image from './Image';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

/**
 * The Image component will:
 *  - render an image within a square box when an image is provided
 *  - render a "pending" placeholder when no image is provided
 *  - render an error when an error state is provided
 */

const ImageFixture = {
  filename: 'IMG_6206.jpeg',
  handle: 'INF7unQm2jNjmirNnY6A',
  mimetype: 'image/jpeg',
  size: 370862,
  status: 'Stored',
  url: 'https://cdn.filestackcontent.com/INF7unQm2jNjmirNnY6A',
};

const ImageUploadFailed = {
  errror: 'Image upload failed',
};

const ImageWrongFormat = {
  errror: 'File provided was an incorrect format',
};

describe('Image Component', () => {
  xtest('Image renders correctly when an image is provided', () => {
    const tree = renderer.create(<Image img={ImageFixture} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Image renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Image img={ImageFixture} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('Image renders correctly when pending', () => {
    const tree = renderer.create(<Image />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  xtest('Image renders correctly when wrong format', () => {
    const tree = renderer.create(<Image />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  xtest('Image renders correctly when upload fails', () => {
    const tree = renderer.create(<Image />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
