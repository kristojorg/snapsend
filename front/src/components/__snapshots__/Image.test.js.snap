// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Component Image renders correctly when pending 1`] = `
<div
  className="sc-bdVaJa ewHNwH"
>
  <div
    className="sc-bwzfXH IQLxr"
  >
    <p
      className="MuiTypography-root-1 MuiTypography-body1-10"
    >
      Uploading image
    </p>
  </div>
</div>
`;

exports[`Image Component Image renders correctly when upload fails 1`] = `
<div
  className="sc-bdVaJa ikyrKG"
>
  uploading image
</div>
`;

exports[`Image Component Image renders correctly when wrong format 1`] = `
<div
  className="sc-bdVaJa ikyrKG"
>
  uploading image
</div>
`;
